### PR TITLE
*: make explicit timestamp on

### PR DESCRIPTION
**Pull request by jackysp**
This pr wants to make the behavior of TiDB consistent with MySQL when it set `explicit_defaults_for_timestamp = on`.
PTAL @zimulala @XuHuaiyu 

MySQL Document: https://dev.mysql.com/doc/refman/5.6/en/server-system-variables.html#sysvar_explicit_defaults_for_timestamp

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -255,6 +255,10 @@ func checkColumnCantHaveDefaultValue(col *table.Column, value interface{}) (err
 	return nil
 }
 
+func isExplicitTimeStamp() bool {
+	return strings.EqualFold(variable.GetSysVar("explicit_defaults_for_timestamp").Value, "ON")
```
Comment (shenli) [inline ddl/ddl_api.go]: GetSysVar will get the default value of the system variable.

Comment (coocood) [reply ddl/ddl_api.go]: This is a None scope variable, it's read-only.

Comment (jackysp) [reply ddl/ddl_api.go]: The code here are used to clarify it should be changed in the future. Anyway, how about return true directly then add a TODO comment on it. @shenli @coocood

**Comment by coocood** [comment]
LGTM

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -255,6 +255,11 @@ func checkColumnCantHaveDefaultValue(col *table.Column, value interface{}) (err
 	return nil
 }
 
+func isExplicitTimeStamp() bool {
```
Comment (XuHuaiyu) [inline ddl/ddl_api.go]: add a comment for this func.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -264,11 +269,14 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
...
 	})
 
-	// Check and set TimestampFlag and OnUpdateNowFlag.
-	if col.Tp == mysql.TypeTimestamp {
-		col.Flag |= mysql.TimestampFlag
-		col.Flag |= mysql.OnUpdateNowFlag
-		col.Flag |= mysql.NotNullFlag
+	isExplicit := isExplicitTimeStamp()
+	if !isExplicit {
+		// Check and set TimestampFlag and OnUpdateNowFlag.
```
Comment (XuHuaiyu) [inline ddl/ddl_api.go]: miss NotNullFlag.

Comment (jackysp) [reply ddl/ddl_api.go]: It should be missed.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -336,7 +344,9 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
 		}
 	}
 
-	setTimestampDefaultValue(col, hasDefaultValue, setOnUpdateNow)
+	if !isExplicit {
```
[the remark this answers is not in the document]
Comment (jackysp) [reply ddl/ddl_api.go]: If we implement explicit off,  this part is useful.

[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl_api.go]: @jackysp 
I mean if you remove this check, the logic is OK. "If isExplicit is false, we don't set the TimestampFlag. So we do nothing in setTimestampDefaultValue."

File: ddl/ddl_api.go (hunk 3 above)
Comment (XuHuaiyu) [reply ddl/ddl_api.go]: But NotNullFlag is set in line281?

**Comment by jackysp** [comment]
PTAL @zimulala

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1266,7 +1276,9 @@ func setDefaultAndComment(ctx sessionctx.Context, col *table.Column, options []*
 		}
 	}
 
-	setTimestampDefaultValue(col, hasDefaultValue, setOnUpdateNow)
+	if !isExplicitTimeStamp() {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could we remove this?

**Comment by jackysp** [comment]
PTAL @zimulala

**Pull request merged**